azure-iot-device/azure/iot/device/iothub/pipeline/mqtt_pipeline.py

```diff
@@ -155,26 +156,66 @@ def _on_disconnected():
             if self.on_disconnected:
                 self.on_disconnected()
 
+        # Set internal event handlers
         self._pipeline.on_pipeline_event_handler = _on_pipeline_event
         self._pipeline.on_connected_handler = _on_connected
         self._pipeline.on_disconnected_handler = _on_disconnected
 
+        # Initialize the pipeline
         callback = EventedCallback()
-
-        # NOTE: It would be nice if this didn't have to go down as a dynamic operation.
-        # At this time we haven't been able to figure out a better way to make it work though.
-
         op = pipeline_ops_base.InitializePipelineOperation(callback=callback)
-
         self._pipeline.run_op(op)
         callback.wait_for_completion()
 
+        # Set the running flag
+        self._running = True
+
+    def _ensure_running(self):
+        if not self._running:
+            raise pipeline_exceptions.PipelineNotRunning(
+                "Cannot execute method - Pipeline is not running"
+            )
+
+    def shutdown(self, callback):
+        """Shut down the pipeline and clean up any resources.
+
+        Once shut down, making any further calls on the pipeline will result in a
+        PipelineNotRunning exception being raised.
+
+        There is currently no way to resume pipeline functionality once shutdown has occurred.
+
+        :raises: :class:`azure.iot.device.iothub.pipeline.exceptions.PipelineNotRunning` if the
+            pipeline has already been shut down
+
+        The shutdown process itself is not expected to fail under any normal condition, but if it
+        does, exceptions are not "raised", but rather, returned via the "error" parameter when
+        invoking "callback".
+        """
+        self._ensure_running()
+        logger.debug("Commencing shutdown of pipeline")
+
+        def on_complete(op, error):
+            if not error:
+                # Only set the pipeline to not be running if the op was successful
```

Comment:
># Only set the pipeline to not be running if the op was successful [](start = 16, length = 66)

So, if you call `shutdown` and it raises an exception, you can still use the pipeline because it's still running?  This doesn't make sense.

Reply:
If there's a failure, the pipeline actually didn't completely shut down. Say there were five pipeline stages, and it failed on stage 3, that means stage 3, 4 and 5 have not been correctly shut down.

I implemented it this way in order to give you the ability to try the operation again so the client can't be put in a state where it is impossible to shutdown.